src/System.Linq/src/System/Linq/Enumerable.cs

```diff
@@ -2181,32 +2420,37 @@ public static TSource Max<TSource>(this IEnumerable<TSource> source)
             TSource value = default(TSource);
             if (value == null)
             {
-                foreach (TSource x in source)
+                using (IEnumerator<TSource> e = source.GetEnumerator())
                 {
-                    if (x != null && (value == null || comparer.Compare(x, value) > 0))
-                        value = x;
+                    if (!e.MoveNext()) return value;
+                    value = e.Current;
+                    // A do…while may seem more natural here, but profiled worse across a wide set of test data
+                    while (value == null)
+                    {
+                        if (!e.MoveNext()) return value;
+                        value = e.Current;
+                    } while (value == null);
+                    while (e.MoveNext())
+                    {
+                        TSource x = e.Current;
+                        if (x != null && comparer.Compare(x, value) > 0) value = x;
+                    }
                 }
-                return value;
             }
             else
             {
-                bool hasValue = false;
-                foreach (TSource x in source)
+                using (IEnumerator<TSource> e = source.GetEnumerator())
                 {
-                    if (hasValue)
-                    {
-                        if (comparer.Compare(x, value) > 0)
-                            value = x;
-                    }
-                    else
+                    if (!e.MoveNext()) throw Error.NoElements();
+                    value = e.Current;
+                    while (e.MoveNext())
                     {
-                        value = x;
-                        hasValue = true;
+                        TSource x = e.Current;
+                        if (comparer.Compare(x, value) > 0) value = x;
                     }
                 }
-                if (hasValue) return value;
-                throw Error.NoElements();
             }
+            return value;
```

Comment:
we could use just one using, and it could be interesting to try 

```C#
                    do
                    {
                        if (!e.MoveNext()) goto exit; // goes to "return value"
                        value = e.Current;
                    }
                    while (value == null);
```